### PR TITLE
Allow passing AWS credentials directly to Amazon S3 and DynamoDB healthcheck

### LINKED ISSUE #482: AWS Health checks assume basic credentials
**What would you like to be added**:
I would like to add a AWSCredentials field to `S3BucketOptions` and `DynamoDBOptions` replacing the current AccessKey and SecretKey string fields (I suggest they become deprecated and still handle that these are specified). These passed AWSCredentials would be used instead of the manual created BasicCredentials to call AWS.

This change could fallback to the default credentials if nothing is specified (thus supporting #119)

**Why is this needed**:

This is useful for cases were basic credentials are not available, for the full list of use cases see the AWS documentation - https://docs.aws.amazon.com/sdk-for-net/v3/developer-guide/net-dg-config-creds.html.

Specifically for us we use AssumeRole credentials as well as AssumeRoleViaWebIdentity credentials to be able to assume IAM roles from our kubernetes service accounts. See https://docs.aws.amazon.com/eks/latest/userguide/specify-service-account-role.html

I am happy to provide a PR that resolves this.

### PR DESCRIPTION
**Pull request by jakubriha**
**What this PR does / why we need it**:
It allows to directly pass `AWSCredentials` object instance to Amazon S3 and DynamoDB healthchecks. This extends use-cases for providing such credentials, like passing [`AssumeRoleAWSCredentials`](https://docs.aws.amazon.com/sdkfornet/v3/apidocs/items/Runtime/TAssumeRoleAWSCredentials.html). This PR was built upon the work of @jocco in his PR #653.

**Which issue(s) this PR fixes**:
close #482

**Special notes for your reviewer**:
No

**Does this PR introduce a user-facing change?**:
The PR makes `AccessKey` and `SecretKey` properties in `S3BucketOptions` and `DynamoDbOptions` obsolete. Instead, it provides new property `Credentials`, which provides the same functionality. However, the obsolete properties are still functioning if the `Credentials` property is not set.

### CODE REVIEW

File: src/HealthChecks.Aws.S3/S3BucketOptions.cs [hunk 1]
```diff
@@ -1,12 +1,17 @@
 ﻿using Amazon.S3;
 using Amazon.S3.Model;
+using Amazon.Runtime;
 using System;
 
 namespace HealthChecks.Aws.S3
 {
     public class S3BucketOptions
     {
+        public AWSCredentials Credentials {get; set;}
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/S3BucketOptions.cs:10]:
```suggestion
        public AWSCredentials Credentials { get; set; }
```

File: src/HealthChecks.Aws.S3/README.md [hunk 2]
```diff
@@ -11,17 +11,49 @@ With all of the following examples, you can additionally add the following param
...
+
+```cs
+public void ConfigureServices(IServiceCollection services)
+{
+    services
+        .AddHealthChecks()
+        .AddS3(options =>
+        {
+            options.BucketName = "bucket-name";
+            options.S3Config = new AmazonS3Config()
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/README.md:53]:
```suggestion
            options.S3Config = new AmazonS3Config
```

File: src/HealthChecks.Aws.S3/S3BucketOptions.cs [hunk 3]
```diff
@@ -1,12 +1,17 @@
...
+
+        [Obsolete("Specify access key and secret as a BasicCredential to AWSCredentials instead")]
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/S3BucketOptions.cs:12]:
```suggestion
        [Obsolete("Specify access key and secret as a BasicAWSCredentials to Credentials property instead")]
```

File: src/HealthChecks.Aws.S3/S3BucketOptions.cs [hunk 4]
```diff
@@ -1,12 +1,17 @@
...
         public string AccessKey { get; set; }
+        [Obsolete("Specify access key and secret as a BasicCredential to AWSCredentials instead")]
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/S3BucketOptions.cs:14]:
```suggestion
        [Obsolete("Specify access key and secret as a BasicAWSCredentials to Credentials property instead")]
```

File: src/HealthChecks.DynamoDb/DynamoDbOptions.cs [hunk 5]
```diff
@@ -1,10 +1,17 @@
...
+using Amazon.Runtime;
 
 namespace HealthChecks.DynamoDb
 {
     public class DynamoDBOptions
     {
-        public string AccessKey { get; set; }
+        public AWSCredentials Credentials { get; set; }
+        
+        [Obsolete("Specify the access key and secret as a BasicCredential to the Credentials Field instead")]
```
Comment (sungam3r) [inline src/HealthChecks.DynamoDb/DynamoDbOptions.cs:11]:
```suggestion
        [Obsolete("Specify the access key and secret as a BasicAWSCredentials to the Credentials property instead")]

```

File: src/HealthChecks.DynamoDb/DynamoDbOptions.cs [hunk 6]
```diff
@@ -1,10 +1,17 @@
...
+        public string AccessKey { get; set;}
+        
+        [Obsolete("Specify the access key and secret as a BasicCredential to the Credentials Field instead")]
```
Comment (sungam3r) [inline src/HealthChecks.DynamoDb/DynamoDbOptions.cs:14]:
```suggestion
        [Obsolete("Specify the access key and secret as a BasicAWSCredentials to the Credentials property instead")]

```

**Comment by jakubriha** [comment]
@sungam3r Thank you for code review suggestions. All of them have been applied.

**Comment by sungam3r** [comment]
@jakubriha Please rebase onto `master`. CI has been changed drastically.

**Comment by sungam3r** [comment]
ping @jakubriha

**Comment by jakubriha** [comment]
@sungam3r Rebase has been done.

File: src/HealthChecks.Aws.S3/README.md [hunk 7]
```diff
@@ -20,8 +20,40 @@ public void ConfigureServices(IServiceCollection services)
...
-            BucketName = "bucket-name",
-            AccessKey = "access-key"
+            options.BucketName = "bucket-name";
+        });
+}
+```
+
+### Directly pass credentials
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/README.md:30]:
```suggestion
```csharp
```

File: src/HealthChecks.Aws.S3/README.md [hunk 8]
```diff
@@ -20,8 +20,40 @@ public void ConfigureServices(IServiceCollection services)
...
+        {
+            options.BucketName = "bucket-name";
+            options.Credentials = new BasicAWSCredentials("access-key", "secret-key");
+        });
+}
+```
+
+### Specify region endpoint
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.Aws.S3/README.md:45]:
```suggestion
```csharp
```

File: src/HealthChecks.DynamoDb/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
+# Amazon DynamoDB Health Check
+
+This health check verifies the ability to communicate with [Amazon DynamoDB](https://aws.amazon.com/dynamodb/).
+
+## Example Usage
+
+With all of the following examples, you can additionally add the following parameters:
+
+- `name`: The health check name. Default if not specified is `aws s3`.
```
Comment (sungam3r) [inline src/HealthChecks.DynamoDb/README.md:9]:
```suggestion
- `name`: The health check name. Default if not specified is `dynamodb`.
```

File: src/HealthChecks.DynamoDb/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+- `failureStatus`: The `HealthStatus` that should be reported when the health check fails. Default is `HealthStatus.Unhealthy`.
+- `tags`: A list of tags that can be used to filter sets of health checks.
+- `timeout`: A `System.TimeSpan` representing the timeout of the check.
+
+### Load credentials from the application's default configuration
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.DynamoDb/README.md:16]:
```suggestion
```csharp
```

File: src/HealthChecks.DynamoDb/README.md [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        .AddDynamoDb(options =>
+        {
+            options.RegionEndpoint = RegionEndpoint.EUCentral1;
+        });
+}
+```
+
+### Directly pass credentials
+
+```cs
```
Comment (sungam3r) [inline src/HealthChecks.DynamoDb/README.md:30]:
```suggestion
```csharp
```

File: test/UnitTests/DependencyInjection/S3/S3UnitTests.cs [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    {
+        [Fact]
+        public void add_health_check_when_properly_configured()
+        {
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddS3(_ => { _.S3Config = new AmazonS3Config(); });
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/S3/S3UnitTests.cs:22]:
```suggestion
            var options = serviceProvider.GetRequiredService<IOptions<HealthCheckServiceOptions>>();
```

File: test/UnitTests/DependencyInjection/S3/S3UnitTests.cs [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+            var registration = options.Value.Registrations.First();
+            var check = registration.Factory(serviceProvider);
+
+            registration.Name.Should().Be("aws s3");
+            check.GetType().Should().Be(typeof(S3HealthCheck));
+        }
+        [Fact]
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/S3/S3UnitTests.cs:30]:
```suggestion

        [Fact]
```

File: test/UnitTests/DependencyInjection/S3/S3UnitTests.cs [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+        public void add_named_health_check_when_properly_configured()
+        {
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddS3(_ => { _.S3Config = new AmazonS3Config(); }, name: "my-s3-group");
+
+            var serviceProvider = services.BuildServiceProvider();
+            var options = serviceProvider.GetService<IOptions<HealthCheckServiceOptions>>();
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/S3/S3UnitTests.cs:38]:
```suggestion
            var options = serviceProvider.GetRequiredService<IOptions<HealthCheckServiceOptions>>();
```

File: test/UnitTests/DependencyInjection/S3/S3UnitTests.cs [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+            check.GetType().Should().Be(typeof(S3HealthCheck));
+        }
+        [Fact]
+        public void add_named_health_check_when_properly_configured()
+        {
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddS3(_ => { _.S3Config = new AmazonS3Config(); }, name: "my-s3-group");
+
+            var serviceProvider = services.BuildServiceProvider();
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/S3/S3UnitTests.cs:37]:
```suggestion
            using var serviceProvider = services.BuildServiceProvider();
```

File: test/UnitTests/DependencyInjection/S3/S3UnitTests.cs [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+    public class s3_registration_should
+    {
+        [Fact]
+        public void add_health_check_when_properly_configured()
+        {
+            var services = new ServiceCollection();
+            services.AddHealthChecks()
+                .AddS3(_ => { _.S3Config = new AmazonS3Config(); });
+
+            var serviceProvider = services.BuildServiceProvider();
```
Comment (sungam3r) [inline test/UnitTests/DependencyInjection/S3/S3UnitTests.cs:21]:
```suggestion
            using var serviceProvider = services.BuildServiceProvider();
```

File: src/HealthChecks.UI/package-lock.json [hunk 17]
```diff
@@ -1,8 +1,5622 @@
 {
```
Comment (sungam3r) [inline src/HealthChecks.UI/package-lock.json:1]: Please revrert all changes in this file.

**Review by sungam3r — CHANGES_REQUESTED** [review]

**Comment by sungam3r** [comment]
I suppose #653 could be closed after merging this one?

**Comment by sungam3r** [comment]
I'm fine to merge afer fixin formatting issues.

**Comment by jakubriha** [comment]
> I suppose #653 could be closed after merging this one?

Yes, it can be closed.

**Comment by sungam3r** [comment]
A bit more work here.

**Comment by sungam3r** [comment]
You still have to update api approval file since `Obsolete` attribute affects public API surface.

**Comment by jakubriha** [comment]
> You still have to update api approval file since `Obsolete` attribute affects public API surface.

Hopefully, the last commit should fix this.

**Review by sungam3r — APPROVED** [review]

**Pull request merged**